src/clj/puppetlabs/services/jruby_pool_manager/impl/reference_pool.clj

```diff
@@ -72,27 +101,19 @@
     (when (jruby-schemas/jruby-instance? instance)
       (let [pool (jruby-internal/get-pool pool-context)
             borrow-count (:borrow-count pool-context)
+            max-borrows (get-in instance [:internal :max-borrows])
             modify-instance-agent (jruby-agents/get-modify-instance-agent pool-context)]
         (.releaseItem pool instance)
         (swap! borrow-count inc)
-        (jruby-agents/send-agent modify-instance-agent
-                                 #(flush-if-at-max-borrows pool-context instance)))))
+        (when (max-borrows-exceeded @borrow-count max-borrows)
+          (jruby-agents/send-agent modify-instance-agent
+                                   #(flush-if-at-max-borrows pool-context instance))))))
 
   (flush-pool
     [pool-context]
     (pool-protocol/lock pool-context)
     (try
-      (let [pool (jruby-internal/get-pool pool-context)
-            borrow-count (:borrow-count pool-context)
-            cleanup-fn (get-in pool-context [:config :lifecycle :cleanup])
-            old-instance (.borrowItem pool)
-            id (inc (:id old-instance))
-            _ (.releaseItem pool old-instance)]
-        ;; This will block waiting for all borrows to be returned
-        (jruby-internal/cleanup-pool-instance! old-instance cleanup-fn)
-        (jruby-agents/add-instance pool-context id)
-        (log/info (i18n/trs "Finished creating JRuby instance with id {0}" id))
-        (reset! borrow-count 0))
+      (flush-pool pool-context)
```

Comment:
How does this know that `flush-pool` doesn't refer to itself, i.e. this function isn't recursive, it's calling the other `flush-pool` above?